File: app/src/main/java/rs/ove/crypt/proto/NativeE2E.java
package rs.ove.crypt.proto;

import android.content.Context;
import android.os.ParcelFileDescriptor;

import java.io.File;
import java.io.IOException;

/** Opaque Android adapter for mst5-client E2E v4. Private keys remain in Rust. */
public final class NativeE2E {
	private NativeE2E() {}

	public static Identity open(Context context, String account, boolean create) throws IOException {
		if (context == null) throw new IOException("Android context is required");
		File directory = new File(context.getFilesDir(), "mst5-e2e");
		if (!directory.isDirectory() && !directory.mkdirs()) throw new IOException("cannot create E2E directory");
		String safe = account == null ? "" : account.replaceAll("[^A-Za-z0-9_.-]", "_");
		if (safe.length() == 0) throw new IOException("E2E account is required");
		File path = new File(directory, safe + ".key");
		long handle = NativeMst5.openE2E(path.getAbsolutePath(), create);
		return new Identity(handle, path.getAbsolutePath());
	}

	public static Identity restore(Context context, String account, String password, Backup backup) throws IOException {
		if (backup == null || backup.version != 2) throw new IOException("invalid E2E backup v2");
		File directory = new File(context.getFilesDir(), "mst5-e2e");
		if (!directory.isDirectory() && !directory.mkdirs()) throw new IOException("cannot create E2E directory");
		String safe = account == null ? "" : account.replaceAll("[^A-Za-z0-9_.-]", "_");
		File path = new File(directory, safe + ".key");
		return new Identity(NativeMst5.restoreE2E(path.getAbsolutePath(), password, backup.encoded), path.getAbsolutePath());
	}

	public static Session session(Identity identity, String peerPublicKey, String from, String to) {
		if (identity == null) throw new IllegalArgumentException("E2E identity is required");
		byte[] peer = Base64Codec.decode(peerPublicKey == null ? "" : peerPublicKey);
		if (peer.length != 32) throw new IllegalArgumentException("E2E public key must be 32 bytes");
		return new Session(identity, peer, from, to);
	}

	public static Envelope seal(Session session, String from, String to, String text) throws IOException {
		byte[] encoded = NativeMst5.e2eSeal(session.identity.requireHandle(), session.peer, from, to,
				(text == null ? "" : text).getBytes("UTF-8"));
		return Envelope.decode(encoded);
	}

	public static String open(Session session, String from, String to, Envelope envelope) throws IOException {
		return new String(NativeMst5.e2eDecrypt(session.identity.requireHandle(), session.peer, from, to, envelope.encoded), "UTF-8");
	}


	/** Exact V2 media-node size: header plus authenticated 64 KiB chunks. */
	public static long encryptedMediaSize(long plaintextSize) throws IOException {
		if (plaintextSize < 0) throw new IOException("invalid E2E media size");
		long chunks = (plaintextSize + 65535L) / 65536L;
		try { return Math.addExact(plaintextSize, Math.addExact(32L, Math.multiplyExact(chunks, 20L))); }
		catch (ArithmeticException error) { throw new IOException("E2E media size overflow", error); }
	}

	public static void uploadMedia(Session session, String endpoint, String publicKey, String ticket,
	                               String fileId, long plaintextSize, ParcelFileDescriptor source,
	                               Mst5MediaClient.Observer observer) throws Exception {
		Mst5MediaClient.uploadE2EDescriptor(endpoint, publicKey, ticket, fileId, plaintextSize, source,
				session.identity.requireHandle(), session.peer, session.from, session.to, observer);
	}

	public static long downloadMedia(Session session, String endpoint, String publicKey, String ticket,
	                                 String fileId, long encryptedSize, ParcelFileDescriptor target,
	                                 Mst5MediaClient.Observer observer) throws Exception {
		return Mst5MediaClient.downloadE2EDescriptor(endpoint, publicKey, ticket, fileId, encryptedSize, target,
				session.identity.requireHandle(), session.peer, session.from, session.to, observer);
	}

	public static byte[] downloadMediaBytes(Session session, String endpoint, String publicKey, String ticket,
	                                        String fileId, long encryptedSize, int maxBytes) throws Exception {
		return Mst5MediaClient.downloadE2EBytes(endpoint, publicKey, ticket, fileId, encryptedSize, maxBytes,
				session.identity.requireHandle(), session.peer, session.from, session.to);
	}

	public static String fingerprint(String publicKey) throws IOException {
		byte[] key = Base64Codec.decode(publicKey == null ? "" : publicKey);
		if (key.length != 32) throw new IllegalArgumentException("E2E public key must be 32 bytes");
		return NativeMst5.e2ePublicFingerprint(key);
	}

	public static Backup backup(Identity identity, String password) throws IOException {
		return new Backup(2, NativeMst5.e2eBackup(identity.requireHandle(), password));
	}

	public static final class Identity {
		private long handle;
		final String path;
		public final String publicKeyB64;
		Identity(long handle, String path) throws IOException {
			this.handle = handle;
			this.path = path;
			this.publicKeyB64 = Base64Codec.encode(NativeMst5.e2ePublicKey(handle));
		}
		public synchronized String fingerprint() throws IOException { return NativeMst5.e2eFingerprint(requireHandle()); }
		public synchronized void close() { NativeMst5.closeE2E(handle); handle = 0; }
		public synchronized void remove() throws IOException { close(); NativeMst5.removeE2E(path); }
		private synchronized long requireHandle() throws IOException {
			if (handle == 0) throw new IOException("E2E identity is closed");
			return handle;
		}
	}

	public static final class Session {
		final Identity identity; final byte[] peer; final String from; final String to;
		Session(Identity identity, byte[] peer, String from, String to) {
			this.identity = identity; this.peer = peer; this.from = from; this.to = to;
		}
	}

	public static final class Envelope {
		public final int version; public final String nonce; public final String ciphertext; public final String tag;
		final byte[] encoded;
		private Envelope(byte[] encoded) {
			this.encoded = encoded; this.version = encoded[0] & 0xff;
			byte[] nonceBytes = new byte[24]; System.arraycopy(encoded, 1, nonceBytes, 0, 24);
			byte[] body = new byte[encoded.length - 25]; System.arraycopy(encoded, 25, body, 0, body.length);
			this.nonce = Base64Codec.encode(nonceBytes); this.ciphertext = Base64Codec.encode(body); this.tag = "";
		}
		public Envelope(int version, String nonce, String ciphertext, String ignoredTag) {
			byte[] n = Base64Codec.decode(nonce); byte[] c = Base64Codec.decode(ciphertext);
			if ((version != 3 && version != 4) || n.length != 24 || c.length < 16) throw new IllegalArgumentException("unsupported E2E envelope");
			this.encoded = new byte[1 + n.length + c.length]; this.encoded[0] = (byte)version;
			System.arraycopy(n, 0, encoded, 1, n.length); System.arraycopy(c, 0, encoded, 1 + n.length, c.length);
			this.version = version; this.nonce = nonce; this.ciphertext = ciphertext; this.tag = "";
		}
		static Envelope decode(byte[] value) throws IOException {
			if (value == null || value.length < 41 || (value[0] != 3 && value[0] != 4)) throw new IOException("unsupported E2E envelope; update the application");
			return new Envelope(value);
		}
	}

	public static final class Backup {
		public final int version; public final byte[] encoded;
		public Backup(int version, byte[] encoded) { this.version = version; this.encoded = encoded; }
	}
}

File: app/src/main/kotlin/rs/ove/crypt/proto/NativeE2E.kt
package rs.ove.crypt.proto

import android.content.Context
import android.os.ParcelFileDescriptor
import java.io.File
import java.io.IOException

/** Opaque Android adapter for mst5-client E2E v4. Private keys remain in Rust. */
object NativeE2E {
    @JvmStatic @Throws(IOException::class) fun open(context: Context?, account: String?, create: Boolean): Identity {
        val root = context ?: throw IOException("Android context is required")
        val directory = File(root.filesDir, "mst5-e2e"); if (!directory.isDirectory && !directory.mkdirs()) throw IOException("cannot create E2E directory")
        val safe = account.orEmpty().replace(Regex("[^A-Za-z0-9_.-]"), "_"); if (safe.isEmpty()) throw IOException("E2E account is required")
        val path = File(directory, "$safe.key").absolutePath; return Identity(NativeMst5.openE2E(path, create), path)
    }
    @JvmStatic @Throws(IOException::class) fun restore(context: Context?, account: String?, password: String?, backup: Backup?): Identity {
        if (backup == null || backup.version != 2) throw IOException("invalid E2E backup v2")
        val root = context ?: throw IOException("Android context is required"); val directory = File(root.filesDir, "mst5-e2e"); if (!directory.isDirectory && !directory.mkdirs()) throw IOException("cannot create E2E directory")
        val safe = account.orEmpty().replace(Regex("[^A-Za-z0-9_.-]"), "_"); val path = File(directory, "$safe.key").absolutePath
        return Identity(NativeMst5.restoreE2E(path, password, backup.encoded), path)
    }
    @JvmStatic fun session(identity: Identity?, peerPublicKey: String?, from: String?, to: String?): Session {
        val actual = identity ?: throw IllegalArgumentException("E2E identity is required"); val peer = Base64Codec.decode(peerPublicKey.orEmpty())
        if (peer.size != 32) throw IllegalArgumentException("E2E public key must be 32 bytes"); return Session(actual, peer, from, to)
    }
    @JvmStatic @Throws(IOException::class) fun seal(session: Session, from: String?, to: String?, text: String?): Envelope = Envelope.decode(NativeMst5.e2eSeal(session.identity.requireHandle(), session.peer, from, to, text.orEmpty().toByteArray(Charsets.UTF_8)))
    @JvmStatic @Throws(IOException::class) fun open(session: Session, from: String?, to: String?, envelope: Envelope): String = NativeMst5.e2eDecrypt(session.identity.requireHandle(), session.peer, from, to, envelope.encoded).toString(Charsets.UTF_8)
    @JvmStatic @Throws(IOException::class) fun encryptedMediaSize(plaintextSize: Long): Long {
        if (plaintextSize < 0) throw IOException("invalid E2E media size")
        return try { Math.addExact(plaintextSize, Math.addExact(32L, Math.multiplyExact((plaintextSize + 65535L) / 65536L, 20L))) } catch (error: ArithmeticException) { throw IOException("E2E media size overflow", error) }
    }
    @JvmStatic @Throws(Exception::class) fun uploadMedia(session: Session, endpoint: String?, publicKey: String?, ticket: String?, fileId: String?, plaintextSize: Long, source: ParcelFileDescriptor?, observer: Mst5MediaClient.Observer?) = Mst5MediaClient.uploadE2EDescriptor(endpoint, publicKey, ticket, fileId, plaintextSize, source, session.identity.requireHandle(), session.peer, session.from, session.to, observer)
    @JvmStatic @Throws(Exception::class) fun downloadMedia(session: Session, endpoint: String?, publicKey: String?, ticket: String?, fileId: String?, encryptedSize: Long, target: ParcelFileDescriptor?, observer: Mst5MediaClient.Observer?): Long = Mst5MediaClient.downloadE2EDescriptor(endpoint, publicKey, ticket, fileId, encryptedSize, target, session.identity.requireHandle(), session.peer, session.from, session.to, observer)
    @JvmStatic @Throws(Exception::class) fun downloadMediaBytes(session: Session, endpoint: String?, publicKey: String?, ticket: String?, fileId: String?, encryptedSize: Long, maxBytes: Int): ByteArray = Mst5MediaClient.downloadE2EBytes(endpoint, publicKey, ticket, fileId, encryptedSize, maxBytes, session.identity.requireHandle(), session.peer, session.from, session.to)
    @JvmStatic @Throws(IOException::class) fun fingerprint(publicKey: String?): String { val key = Base64Codec.decode(publicKey.orEmpty()); if (key.size != 32) throw IllegalArgumentException("E2E public key must be 32 bytes"); return NativeMst5.e2ePublicFingerprint(key) }
    @JvmStatic @Throws(IOException::class) fun backup(identity: Identity, password: String?): Backup = Backup(2, NativeMst5.e2eBackup(identity.requireHandle(), password))

    class Identity internal constructor(private var handle: Long, @JvmField internal val path: String) {
        @JvmField val publicKeyB64: String = Base64Codec.encode(NativeMst5.e2ePublicKey(handle))
        @Synchronized @Throws(IOException::class) fun fingerprint(): String = NativeMst5.e2eFingerprint(requireHandle())
        @Synchronized fun close() { NativeMst5.closeE2E(handle); handle = 0 }
        @Synchronized @Throws(IOException::class) fun remove() { close(); NativeMst5.removeE2E(path) }
        @Synchronized @Throws(IOException::class) internal fun requireHandle(): Long { if (handle == 0L) throw IOException("E2E identity is closed"); return handle }
    }
    class Session internal constructor(@JvmField internal val identity: Identity, @JvmField internal val peer: ByteArray, @JvmField internal val from: String?, @JvmField internal val to: String?)
    class Envelope private constructor(@JvmField internal val encoded: ByteArray) {
        @JvmField val version: Int = encoded[0].toInt() and 0xff
        @JvmField val nonce: String = Base64Codec.encode(encoded.copyOfRange(1, 25))
        @JvmField val ciphertext: String = Base64Codec.encode(encoded.copyOfRange(25, encoded.size))
        @JvmField val tag: String = ""
        constructor(version: Int, nonce: String?, ciphertext: String?, ignoredTag: String?) : this(composeEnvelope(version, nonce, ciphertext))
        companion object {
            private fun composeEnvelope(version: Int, nonce: String?, ciphertext: String?): ByteArray {
                val n = Base64Codec.decode(nonce.orEmpty()); val c = Base64Codec.decode(ciphertext.orEmpty())
                if ((version != 3 && version != 4) || n.size != 24 || c.size < 16) throw IllegalArgumentException("unsupported E2E envelope")
                return ByteArray(1 + n.size + c.size).also { out -> out[0] = version.toByte(); n.copyInto(out, 1); c.copyInto(out, 1 + n.size) }
            }
            @JvmStatic @Throws(IOException::class) fun decode(value: ByteArray?): Envelope { if (value == null || value.size < 41 || (value[0].toInt() != 3 && value[0].toInt() != 4)) throw IOException("unsupported E2E envelope; update the application"); return Envelope(value) }
        }
    }
    class Backup(@JvmField val version: Int, @JvmField val encoded: ByteArray?)
}
